pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+            )
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        self.nqubits = max(circuit.device.wires) + 1
+
+        if self.nqubits > 4:
+            print(
+                "WARNING: The exact Lie gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow."
+            )
+
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators()
+
+        if not isinstance(hamiltonian, qml.Hamiltonian):
+            raise TypeError(
+                f"`hamiltonian` must be a `qml.Hamiltonian`, "
+                f"received {type(hamiltonian)} "
+            )
+
+        self.hamiltonian = hamiltonian
+        self.coeffs, self.observables = self.hamiltonian.terms
+        pl_pauli_to_idx = {"PauliX": 1, "PauliY": 2, "PauliZ": 3}
+        hamiltonian_basis_names = []
+        for op in hamiltonian.ops:
+            name = [0] * self.nqubits
+            if len(op.wires) > 1:
+                for loc, p in zip(op.wires, op.name):
+                    name[loc] = pl_pauli_to_idx[p]
+            else:
+                name[op.wires[0]] = pl_pauli_to_idx[op.name]
+            hamiltonian_basis_names.append(tuple(name))
+        self.stepsize = stepsize
+
+        @batch_transform
+        def algebra_commutator(tape, observables):
+            """Calculate the Lie gradient with the parameter shift rule (see `get_omegas`)"""
+            tapes = []
+            for obs in observables:
+                tapes_plus = [
+                    JacobianTape(p + "_p") for p in self.lie_algebra_basis_names
+                ]
+                tapes_min = [
+                    JacobianTape(p + "_m") for p in self.lie_algebra_basis_names
+                ]
+
+                # loop through all operations on the input tape
+                for op in tape.operations:
+                    for t in tapes_plus + tapes_min:
+                        with t:
+                            qml.apply(op)
+                for i, t in enumerate(tapes_plus):
+                    with t:
+                        qml.PauliRot(
+                            np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                for i, t in enumerate(tapes_min):
+                    with t:
+                        qml.PauliRot(
+                            -np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                tapes.append((tapes_plus, tapes_min))
+            return tapes, None
```

Comment:
just to clarify, this creates `2 * N_L * N_o` tapes, where `N_L` is the size of the Lie algebra, and `N_o` is the number of observables in the Hamiltonian?

Reply:
I'm curious if you can reduce the number of circuits needed by grouping the Hamiltonian into qubit-wise commuting terms?

Reply:
nevermind, I see that you are already grouping!

Reply:
Yes, we create `2 * N_L * N_o` tapes.